lib/galaxy/authnz/custos_authnz.py

```diff
@@ -475,3 +460,118 @@ def _username_from_userinfo(self, trans, userinfo):
             return f"{username}{count}"
         else:
             return username
+
+
+class OIDCAuthnzBaseKeycloak(OIDCAuthnzBase):
+    def __init__(self, provider, oidc_config, oidc_backend_config, idphint=None):
+        super().__init__(provider, oidc_config, oidc_backend_config, idphint)
+        self.config.extra_params = {
+            "kc_idp_hint": oidc_backend_config.get("idphint", "oidc")
+        }
+        self._load_config()
+
+    def _get_provider_specific_scopes(self):
+        return ["org.cilogon.userinfo"]
+
+
+class OIDCAuthnzBaseCiLogon(OIDCAuthnzBase):
+    def __init__(self, provider, oidc_config, oidc_backend_config, idphint=None):
+        super().__init__(provider, oidc_config, oidc_backend_config, idphint)
+        self.config.extra_params = {
+            "kc_idp_hint": oidc_backend_config.get("idphint", "cilogon")
+        }
+        self._load_config()
+
+    def _get_provider_specific_scopes(self):
+        return ["org.cilogon.userinfo"]
+
+    def _get_well_known_uri_from_url(self, provider):
+        base_url = self.config.url
+        # backwards compatibility. CILogon URL is given with /authorize in the examples. not sure if
+        # this applies to the wild, but let's be safe here and remove the /authorize if it exists
+        # which will lead to the correct openid configuration
+        base_url = base_url if base_url.split("/")[-1] != "authorize" else "/".join(base_url.split("/")[:-1])
+        return f"{base_url}/.well-known/openid-configuration"
+
+
+class CustosAuthFactory:
+    @dataclass
+    class _CustosAuthBasedProviderCacheItem:
+        created_at : datetime
+        item: OIDCAuthnzBase
+        provider: str
+        oidc_config: dict
+        oidc_backend_config: dict
+        idphint: str
+
+    _CustosAuthBasedProvidersCache: List[_CustosAuthBasedProviderCacheItem] = []
+
+    @staticmethod
+    def GetCustosBasedAuthProvider (provider, oidc_config, oidc_backend_config, idphint=None):
+        # remove old entries (older than 15 minutes)
+        _CustosAuthBasedProvidersCache = filter(lambda x: datetime.now() - x.created_at < timedelta(minutes=15),
```

Comment:
Is this to cache well-known info? If so, I think it changes rarely enough, if ever, that we can afford a restart if that info changes, so my personal preference would be to simplify further by removing the cache.

Reply:
Yes, this will cache the well known config for 15 minutes. If you don't want the cache to expire, just remove line 512. However having a cache in here makes galaxy 'repair' itself after 15 minutes of a oidc well-known change. 

Not caching at all is IMHO a bad idea. Due to the design of the code the well-known is fetched every request to galaxy (by a user with an OIDC account). As a result every request to galaxy causes a request to the .well-known... That should cause quite some server load and also stall the transaction while the request of .well-known is in progress.  
We could further refactor to only fetch the .well-known upon refresh/authenticate/callback but I believe that is a more fragile design than simply caching it.

Reply:
I'm not suggesting reloading on every request. I'm saying load it once permanently (using a module level variable), and getting rid of 15 minute caching. Afaik, OIDC well known info does not change, and on the remote chance that it does, an admin can just trigger a process level restart.

Reply:
To put it differently, if cache expiry has minimal benefit, it's better to remove it and simplify. If it does serve a purpose, it would be better to use a library like TTLCache from cachetools.

Reply:
removed the TTL